### ISSUE TITLE
Convolutional-LSTM models for object Detection on Video with much higher accuracy

### ISSUE THREAD

**Issue by AlexeyAB** [opencv/opencv#15033]
##### System information (version)

- OpenCV =>  4.1.0 - master-branch
- Operating System / Platform => Win/Linux
- Compiler => g++ 5.4.0

##### Detailed description

Required implementation of Convolutional-LSTM layer which seems to give a huge increase in accuracy for Detection on Video: https://github.com/AlexeyAB/darknet/issues/3114

It should allow to use this **Yolov3-SPP-LSTM** model with new layer `[conv-lstm]` for object detection on Video
* cfg: https://github.com/AlexeyAB/darknet/files/3199654/yolo_v3_spp_lstm.cfg.txt
* weights: https://drive.google.com/open?id=1-lFybALcypqTRtT3pzrrVuCRtDeWjsog

```
[conv_lstm]
batch_normalize=1
size=3
pad=1
output=256
peephole=0
state_constrain=512
activation=leaky
```

Implementation is got from this paper - page 4: https://arxiv.org/pdf/1506.04214v1.pdf

Just:
* `peephole=0` - disables Peephole `W * C`
* `peephole=1` - enables Peephole, but Peephole uses `*` - Convolution instead of `o` - Element-wise-product (Hadamard product), so the weights-file doesn't depend on network resolution (we can change width= and height= in [net]-section without retraining the model)

* `state_constrain=512` - the maximum absolute value of `C`-state: https://github.com/AlexeyAB/darknet/blob/5a6afe96d3aa8aed19405577db7dba0ff173c848/src/conv_lstm_layer.c#L523 By default `state_constrain = time_steps * 32` (time_steps param from cfg-file that is used for training only, but state_constrain is used for both Training and Detection)

There is the same proposal for Pytorch: https://github.com/pytorch/pytorch/issues/1706

![58437703-cc90b080-80d3-11e9-9aa1-e2451c601398](https://user-images.githubusercontent.com/4096485/61124814-9e630680-a4b0-11e9-9fce-042832210fff.png)

With this implementation: https://github.com/AlexeyAB/darknet/blob/5a6afe96d3aa8aed19405577db7dba0ff173c848/src/conv_lstm_layer.c#L852-L929

##### Steps to reproduce

You can run any of these models `cfg/weights`: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968
by using this repository: https://github.com/AlexeyAB/darknet
and this command:
`./darknet detector demo data/self_driving.data yolov3-tiny_occlusion_track.cfg yolov3-tiny_occlusion_track_last.weights self_driving_valid.avi -out_filename out_self_driving_valid.avi`

* `self_driving.data` 
[self_driving.data.txt](https://github.com/opencv/opencv/files/3386250/self_driving.data.txt)

* `self_driving_valid.avi` file: https://drive.google.com/open?id=1rdxV1hYSQs6MNxBSIO9dNkAiBvb07aun

**Comment by Scolymus** [comment]
Any news on it?

**Comment by pushkalkatara** [comment]
I noticed that there is an implementation of the convolution operation in the [conv_layer.cpp](https://github.com/opencv/opencv/blob/3133bb49aa1af0152b155d6299c3fcecefce317a/modules/dnn/src/layers/convolution_layer.cpp#L607) and the element-wise multiplication can be done using A.mul(B). I am trying to implement the layer, where should I start the implementation? [recurrent_layers.cpp] or [conv_layer.cpp] or a new file?

**Comment by Scolymus** [comment]
@pushkalkatara were you able to implement it?

**Comment by scianand** [comment]
Hi @alalek, I just wanted to know, is there any way to implement conv_lstm layer in cv::dnn. I have trained my conv_lstm model with AlexeyAB darknet and there is currently no support for conv_lstm layer in cv::dnn. Please, can you help me out?

Thanks,

**Comment by vpisarev** [comment]
Hello, @AlexeyAB, @berak, @dkurt, looks to me like a good project for GSoC and/or Intel Summer school. Do some of you guys want to be mentors? Or maybe have good students who can implement it? If you like this idea, can you shape it as GSoC idea that can be added herehttps://github.com/opencv/opencv/wiki/GSoC_2020?

**Comment by Eclipsiess** [comment]
Hello @AlexeyAB  @vpisarev  @asmorkalov is this issue still available for contribution ? I want to work on this issue   .  I have worked on lstm as well as image detection  models and I would like to contribute to this . I noticed that this  has been added to Gsoc label  .

**Comment by Wodlfvllf** [comment]
I participated as a contributor last year in gsoc 2024. I would like to contribute to implementing the Convolutional-LSTM layer to enhance video object detection accuracy. I have hands-on experience with PyTorch, TensorFlow, and advanced machine learning concepts like RNNs and Transformers. I aim to align the implementation with OpenCV's coding standards while improving computational efficiency. Let me know how I can proceed!
